### ISSUE TITLE
Impossible to test puppet5 modules with eyaml data sources?

### ISSUE THREAD

**Issue by LongLiveCHIEF** [voxpupuli/hiera-eyaml#253]
I've been struggling for days trying to test a new puppet 5 module that contains a hiera-eyaml data source. It seems that due to the nature of how hiera's layer hierarchy works, it's nearly impossible to prevent eyaml from decrypting a value during unit testing.

I'm using [pdk](https://github.com/puppetlabs/pdk), (which uses rspec-puppet) and here's what i've tried and the failures i've encountered along the way:

### No .pem key files

My first error when running unit tests was a `error @rb_open file does not exists /etc/puppetlabs/secure/keys/key.pem`. 

To resolve this, I tried:

- providing a default value for the two class params that have eyaml data-sources in my hierarchy. 
- overriding the `EYAML_CONFIG` location, with config file defining different locations for key files (and creating fake keys)
- setting `hiera_config` value to an alternative hiera.yaml in fixutres, that didn't contain any eyaml data sources

both of these still resulted in the error above. It seems that since the key locations specified in the module's hiera.yaml file are now as of puppet 4.8/5 part of hiera layer 3, they override any value/location passed using a config file defined using the `EYAML_CONFIG` environment variable, or even 

Next, I tried:

- providing fake keys for test environments, and creating a level of higher specificity in hierarchy that essentially allowed me to tac on `-testing` to any eyaml source file to use the test keys and values in stead of the encrypted values that require the private key stored on the puppetserver to decrypt.

Even though the hierarchy was no longer using the real encrypted data sources in APL, I now get errors for a `bad decrypt`:

```
failed: rspec: ./spec/classes/some_profile_spec.rb:26: error during compilation: Evaluation Error: Error while evaluating a Resource Statement, bad decrypt at line 2:1 on node a43bf5579874
```

I've tried just about everything, but it seems that due to the nature of specificity for layer 3 data-sources, the old ways of tricking the unit test suite to use test values no longer work.

Is there a known way/documenation/examples to unit test puppet 5 modules with eyaml data sources? A way to mock the eyaml lookup, or override the config provided by the layer 3 `hiera.yaml` file?

This relates to #252, since I believe if I could use variable interpolation in my layer 3 `hiera.yaml` file, I could easily use testing keys for testing to get rid of the `bad decrypt` errors in my unit testing suite.

**Comment by LongLiveCHIEF** [comment]
Able to reproduce: https://github.com/LongLiveCHIEF/eyaml-rspec-example-failure/blob/add-eyaml-data-source/data/sector/alpha.yaml#L5

Looks like you don't wind up with any of the `bad decrypt` errors, until you try to use `lookup` from within a data source. 

To reproduce, I made it convenient by wrapping everything with docker and make:

```
git clone -b add-eyaml-data-source https://github.com/LongLiveCHIEF/eyaml-rspec-example-failure.git
cd eyaml-rspec-example-failure
make build
make test
```

**Comment by jkwmoore** [comment]
FYI - I cannot claim to be a Puppet backend expert but I managed to get past PDK testing issues around this as below.

---

### What was the problem?

PDK tests are failing because I don't have the keys necessary to decrypt the secrets while testing locally. 
I also do not want to keep the encryption keys locally, nor anywhere except the PuppetServer.

So what do these errors look like?

Well, I am trying to use the following from my hiera ``common.eyaml``:

```yaml
myclass::mysecretstring: ENC[PKCS7,MYENCRYPTEDSECRETSTRINGCONTENT]
```

And PDK gives me:

```shell
  1) myclass::mymanifest on redhat-8-x86_64 is expected to compile into a catalogue without dependency cycles
     Failure/Error: it { is_expected.to compile }
       error during compilation: Function lookup() did not find a value for the name 'myclass::mysecretstring' on node my_dev_machine.local
     # ./spec/classes/mymanifest_spec.rb:10:in `block (4 levels) in <top (required)>'
```

### How did I sort this?

I added overrides to the spec test used for the class.

```ruby
      before do
        allow(Puppet::Pops::Lookup).to receive(:lookup).and_call_original  # Allow normal lookups

        allow(Puppet::Pops::Lookup).to receive(:lookup)
          .with('myclass::mysecretstring', anything, anything, anything, anything, anything)
          .and_return('MOCKED_VALUE_STRING')
      end
```

### Why do I think this works

Aside of the fact I tested it and it appears to do what I wanted, see below:


#### **What is `Puppet::Pops::Lookup`?**
- Puppet uses **Hiera** to retrieve values from a hierarchy of data sources.
- When a class requests a variable via `lookup('some_variable')`, Puppet internally calls `Puppet::Pops::Lookup.lookup('some_variable', ...)`.
- This function searches for the requested key across the defined Hiera hierarchy.
- https://www.rubydoc.info/gems/puppet/Puppet%2FPops%2FLookup.lookup

For example, if the Puppet manifest has:
```puppet
$mysecret = lookup('myclass::mysecretstring')
```
Puppet will call:
```ruby
Puppet::Pops::Lookup.lookup('myclass::mysecretstring', ...)
```
and return the corresponding value from Hiera.

---

#### Stub Only the Problematic Lookup
```ruby
allow(Puppet::Pops::Lookup).to receive(:lookup)
  .with('myclass::mysecretstring', anything, anything, anything, anything, anything)
  .and_return('MOCKED_VALUE_STRING')
```
- This overrides lookups **only** for `myclass::mysecretstring`.
- `.with(...)` ensures that this stub only activates when the exact expected arguments are passed.
- `.and_return('MOCKED_VALUE_STRING')` means that **whenever Puppet tries to look up this variable, it will get `'MOCKED_VALUE_STRING'` instead of failing.**
- Depending on the expected data type for encrypted values, different mocking values may be required.

**Effect:**  
- `lookup('myclass::mysecretstring')` **returns `'MOCKED_VALUE_STRING'` instead of failing**.  
- All other lookups still use Hiera data lookups as normal/expected.

---

#### Allow Normal Lookups
```ruby
allow(Puppet::Pops::Lookup).to receive(:lookup).and_call_original
```
- `allow(...).to receive(:lookup)` tells RSpec to intercept calls to `Puppet::Pops::Lookup.lookup`.
- `.and_call_original` ensures that all other lookups (except the one we explicitly stub) continue to function normally.
- Without this, RSpec would try to stub all lookups, causing unexpected failures when other values like `myclass::someotherstring'` are requested from ``common.yaml``.

**Effect:**  
- Any `lookup('some_variable')` that is not explicitly stubbed will still perform its normal function.

---

### **tl;dr Why I think this works and warnings**
1. **Ensures normal lookups proceed without modification** → The `.and_call_original` call ensures we do not break unrelated lookups.
2. **Only stubs the problematic keys** → This avoids unnecessary stubbing and allows normal Puppet behavior.
3. **Prevents lookup errors in tests** → PDK tests no longer fail due to "missing" Hiera data.
4. Depending on the expected data type for encrypted values, different mocking values may be required.
5. I am not a Ruby expert nor Puppet/PDK backend expert - here be dragons!

GLHF!
JKWM.
